### PR TITLE
AsyncRestTemplate instrumentation

### LINKED ISSUE #6: Instrument AsyncRestTemplate
Instrument AsyncRestTemplate [1] and provide auto-configuration.

[1]: javadoc http://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/client/AsyncRestTemplate.html

### PR DESCRIPTION
**Pull request by pavolloffay**
resolves #6

### CODE REVIEW

File: opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingAsyncRestTemplateTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.util.concurrent.ListenableFuture;
+import org.springframework.web.client.AsyncRestTemplate;
+
+import java.util.Collections;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * @author Pavol Loffay
+ */
+public class TracingAsyncRestTemplateTest extends AbstractTracingClientTest<AsyncRestTemplate> {
```
Comment (jpkrohling) [inline opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingAsyncRestTemplateTest.java]: You need far more tests for the async template than for the regular sync one :) Please, add more tests that exercise the async features.

File: opentracing-spring-web-autoconfigure/src/test/java/io/opentracing/contrib/spring/web/autoconfig/AsyncRestTemplateAutoConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Autowired
+    private MockTracer mockTracer;
+
+    @Autowired
+    @Qualifier("bar")
+    private AsyncRestTemplate asyncRestTemplate;
+
+    @Test
+    public void testTracingAsyncRequest() throws ExecutionException, InterruptedException {
```
Comment (objectiser) [inline opentracing-spring-web-autoconfigure/src/test/java/io/opentracing/contrib/spring/web/autoconfig/AsyncRestTemplateAutoConfigurationTest.java]: Agree more tests are required - also checking when an existing parent span has been created.

File: opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/TracingAsyncRestTemplateInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            try {
+                spanDecorator.onRequest(httpRequest, span);
+            } catch (RuntimeException exDecorator) {
+                log.log(Level.SEVERE, "Exception during decorating span", exDecorator);
+            }
+        }
+
+        ListenableFuture<ClientHttpResponse> future = execution.executeAsync(httpRequest, body);
+
+        future.addCallback(new ListenableFutureCallback<ClientHttpResponse>() {
```
Comment (objectiser) [inline opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/TracingAsyncRestTemplateInterceptor.java]: Not sure if I've misunderstood, but this shouldn't be an issue. How could the callback be added before the future is returned from the executeAsync?

**Comment by pavolloffay** [comment]
Added a test which simultaneously fires multiple requests

File: opentracing-spring-web-autoconfigure/src/test/java/io/opentracing/contrib/spring/web/autoconfig/AsyncRestTemplateAutoConfigurationTest.java (hunk 2 above)
Comment (pavolloffay) [inline opentracing-spring-web-autoconfigure/src/test/java/io/opentracing/contrib/spring/web/autoconfig/AsyncRestTemplateAutoConfigurationTest.java]: see https://github.com/opentracing-contrib/java-spring-web/pull/17/commits/3be48a880fb091ffd32d2b965ddec47da80d74ba

File: opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingAsyncRestTemplateTest.java [hunk 4]
```diff
@@ -40,4 +50,42 @@ public AsyncRestTemplate template() {
 
         mockServer = MockRestServiceServer.bindTo(client.template()).build();
     }
+
+    @Test
+    public void testMultipleRequests() throws InterruptedException {
```
Comment (jpkrohling) [inline opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingAsyncRestTemplateTest.java]: I don't think this test is exercising the tricky parts of an async setup :-) Ideally, you'd ensure that data didn't leak from one thread to another. 

For instance, you could have 5 calls to each of those:

`http://localhost:8080/foo/span1`, `http://localhost:8080/foo/span2`, `http://localhost:8080/foo/span3`. The parent span for `/foo/span1` is `span1` and you'd check afterwards if the parent span you got on `span1` was seen as `span1`. This way, you are testing that your data is protected even in the case of concurrent async requests.

Comment (pavolloffay) [inline opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingAsyncRestTemplateTest.java]: I'm not sure what exactly you are requesting.

Test executes multiple async requests and tests if each span representing a given request is linked with a parent span. 


Edit: Now test creates a parent span for each async reques.

File: opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingAsyncRestTemplateTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+        executorService.awaitTermination(1, TimeUnit.SECONDS);
+        executorService.shutdown();
+
+        List<MockSpan> mockSpans = mockTracer.finishedSpans();
+        Assert.assertEquals(numberOfCalls, mockSpans.size());
+
+        for (int i = 0; i < numberOfCalls; i++) {
+            MockSpan mockSpan = mockSpans.get(0);
```
Comment (objectiser) [inline opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingAsyncRestTemplateTest.java]: shoudn't this be `get(i)`?

Comment (pavolloffay) [inline opentracing-spring-web/src/test/java/io/opentracing/contrib/spring/web/client/TracingAsyncRestTemplateTest.java]: wow :+1: , thanks

**Comment by pavolloffay** [comment]
@jpkrohling @objectiser thanks!

**Pull request merged**